wolfcrypt/src/rsa.c

```diff
@@ -1000,374 +1034,472 @@ static int wc_RsaFunction(const byte* in, word32 inLen, byte* out,
-    return sz;
+#ifdef WOLFSSL_ASYNC_CRYPT
+    if (key->asyncDev.marker == WOLFSSL_ASYNC_MARKER_RSA) {
+        ret = wc_RsaFunctionAsync(in, inLen, out, outLen, type, key, rng);
+    }
+    else
+#endif
+    {
+        ret = wc_RsaFunctionSync(in, inLen, out, outLen, type, key, rng);
+    }
+
+    if (ret == MP_EXPTMOD_E) {
+        /* This can happen due to incorrectly set FP_MAX_BITS or missing XREALLOC */
+        WOLFSSL_MSG("RSA_FUNCTION MP_EXPTMOD_E: memory/config problem");
+    }
+    return ret;
 }
 
 
-#ifndef WC_NO_RSA_OAEP
+/* Internal Wrappers */
 /* Gives the option of choosing padding type
    in : input to be encrypted
    inLen: length of input buffer
    out: encrypted output
    outLen: length of encrypted output buffer
    key   : wolfSSL initialized RSA key struct
    rng   : wolfSSL initialized random number struct
-   type  : type of padding to use ie WC_RSA_OAEP_PAD
+   rsa_type  : type of RSA: RSA_PUBLIC_ENCRYPT, RSA_PUBLIC_DECRYPT, 
+        RSA_PRIVATE_ENCRYPT or RSA_PRIVATE_DECRYPT
+   pad_value: RSA_BLOCK_TYPE_1 or RSA_BLOCK_TYPE_2
+   pad_type  : type of padding: WC_RSA_PKCSV15_PAD or  WC_RSA_OAEP_PAD
    hash  : type of hash algorithm to use found in wolfssl/wolfcrypt/hash.h
    mgf   : type of mask generation function to use
    label : optional label
    labelSz : size of optional label buffer */
-int wc_RsaPublicEncrypt_ex(const byte* in, word32 inLen, byte* out,
-                    word32 outLen, RsaKey* key, WC_RNG* rng, int type,
-                    enum wc_HashType hash, int mgf, byte* label, word32 labelSz)
+static int RsaPublicEncryptEx(const byte* in, word32 inLen, byte* out,
+                            word32 outLen, RsaKey* key, int rsa_type,
+                            byte pad_value, int pad_type,
+                            enum wc_HashType hash, int mgf,
+                            byte* label, word32 labelSz, WC_RNG* rng)
 {
-    int sz, ret;
+    int ret = BAD_FUNC_ARG, sz;
 
-#ifdef HAVE_CAVIUM
-    if (key->magic == WOLFSSL_RSA_CAVIUM_MAGIC)
-        return CaviumRsaPublicEncrypt(in, inLen, out, outLen, key);
-#endif
+    if (in == NULL || inLen == 0 || out == NULL || key == NULL) {
+        return ret;
+    }
 
-    sz = mp_unsigned_bin_size(&key->n);
-    if (sz > (int)outLen)
+    sz = wc_RsaEncryptSize(key);
+    if (sz > (int)outLen) {
         return RSA_BUFFER_E;
+    }
 
     if (sz < RSA_MIN_PAD_SZ) {
         return WC_KEY_SIZE_E;
     }
 
-    if (inLen > (word32)(sz - RSA_MIN_PAD_SZ))
+    if (inLen > (word32)(sz - RSA_MIN_PAD_SZ)) {
         return RSA_BUFFER_E;
+    }
 
-    ret = wc_RsaPad_ex(in, inLen, out, sz, RSA_BLOCK_TYPE_2, rng,
-                                    type, hash, mgf, label, labelSz, key->heap);
-    if (ret != 0)
-        return ret;
+    /* Optional key type check (disabled by default) */
+    /* Note: internal tests allow private to be used as public */
+#ifdef RSA_CHECK_KEYTYPE
+    if ((rsa_type == RSA_PUBLIC_ENCRYPT && key->type != RSA_PUBLIC) || 
```

Comment:
A private key is a public key, i.e., a private key can do public ops.  Not just for internal testing.